demos/bbm_auxiliary/demo_bbm_aux.py.rst

```diff
@@ -0,0 +1,208 @@
+The time discretization puts the main unknown in a continuous space and the
+auxiliary variable in a discontinuous one.  See equation (7.17) of Andrews'
+thesis for the particular formulation.
+
+
+Firedrake imports::
+
+  from firedrake import (
+      Constant, Function, FunctionSpace,
+      PeriodicIntervalMesh, SpatialCoordinate, TestFunction, TestFunctions,
+      assemble, dx, errornorm, exp, grad, inner, interpolate, norm, project,
+      solve, split
+  )
+
+  from FIAT import make_quadrature, ufc_simplex
+  from FIAT.quadrature_schemes import create_quadrature
+
+  from irksome import Dt, GalerkinTimeStepper
+  from irksome.labeling import TimeQuadratureLabel
+
+  def sech(x):
+      return 2 / (exp(x) + exp(-x))
+
+
+  N = 8000
+  L = 100
+  h = L / N
+  msh = PeriodicIntervalMesh(N, L)
+
+  c = Constant(0.5)
+
+  t = Constant(0)
+  dt = Constant(10*h)
+
+  x, = SpatialCoordinate(msh)
+
+  center = 40.0
+  delta = -c * center
+
+  uexact = 3 * c**2 / (1-c**2) \
+      * sech(0.5 * (c * x - c * t / (1 - c ** 2) + delta))**2
+
+  space_deg = 3
+  time_deg = 1
+
+This sets up the function space for the unknown :math:`u` and
+auxiliary variable :math:`\tilde{wH}`::
+
+  V = FunctionSpace(msh, "Hermite", space_deg)
+  Z = V * V
+
+  uwHtilde = Function(Z)
+  uwHtilde.subfunctions[0].project(uexact)
+
+  wHtilde = Function(V)
+
+We need a consistent initial condition for :math:`\tilde{wH}`::
+  
+  v = TestFunction(V)
+  Finit = (inner(wHtilde, v) * dx
+           + inner(wHtilde.dx(0), v.dx(0)) * dx
+           - inner(uexact + 0.5 * uexact**2, v) * dx)
+  solve(Finit==0, wHtilde)
+  uwHtilde.subfunctions[1].assign(wHtilde)
+
+  u, wHtilde = split(uwHtilde)
+
+Output the initial condition to disk::
+
+  xs = msh.coordinates.dat.data[::2]
+
+  with open("bbm_aux_init.csv", "w") as outfile:
+      outfile.write("x,u\n")
+      for xcur, ucur in zip(xs, uwHtilde.subfunctions[0].dat.data):
+          outfile.write("%f,%f\n" % (xcur, ucur))
+
+  v, vH = TestFunctions(Z)
+
+Create temporal quadrature rules in FIAT::
+  
+  ufc_line = ufc_simplex(1)
+  time_order_low = 2* time_deg
+  time_order_high = 20
+  Qlow = create_quadrature(ufc_line, time_order_low)
+  Qhigh = create_quadrature(ufc_line, time_order_high)
+
+
+  Llow = TimeQuadratureLabel(Qlow.get_points(), Qlow.get_weights())
+  Lhigh = TimeQuadratureLabel(Qhigh.get_points(), Qhigh.get_weights())
+
+
+  def h1inner(u, v):
+      return inner(u, v) + inner(grad(u), grad(v))
+
+
+This tags several of the terms with a low-order time integration scheme,
+but forces a higher-order method on the nonlinear term::
+
+  F = Llow(h1inner(Dt(u), v) * dx
+         - 0.5 * h1inner(wHtilde, v.dx(0)) * dx
+         + 0.5 * h1inner(wHtilde.dx(0), v) * dx
+         + h1inner(wHtilde, vH) * dx) \
+         - Lhigh(inner(u + 0.5 * u**2, vH) * dx)
+
+
+  params = {"mat_type": "aij",
```

Comment:
Are any of these options strictly necessary?

Reply:
At one point (when I had the wrong variational form, it turns out) I turned on solver parameters.  But I just checked and it runs fine default parameters.  So I've removed them.